src/world/Actor/BNpc.cpp

```diff
@@ -151,13 +154,85 @@ void Sapphire::Entity::BNpc::setState( BNpcState state )
   m_state = state;
 }
 
+void Sapphire::Entity::BNpc::step()
+{
+  if( m_naviLastPath.empty() )
+    // No path to track
+    return;
+
+  if( Util::distance( getPos().x, getPos().y, getPos().z, m_naviTarget.x, m_naviTarget.y, m_naviTarget.z ) <= 4 )
+  {
+    // Reached target
+    m_naviLastPath.clear();
+    return;
+  }
+
+  auto stepPos = m_naviLastPath[m_naviPathStep];
+
+  if( Util::distance( getPos().x, getPos().y, getPos().z, stepPos.x, stepPos.y, stepPos.z ) <= 4 && m_naviPathStep < m_naviLastPath.size() - 1 )
+  {
+    // Reached step in path
+    m_naviPathStep++;
+
+    stepPos = m_naviLastPath[m_naviPathStep];
+  }
+
+  // This is probably not a good way to do it but works fine for now
+  float rot = Util::calcAngFrom( getPos().x, getPos().z, stepPos.x, stepPos.z );
+  float newRot = PI - rot + ( PI / 2 );
+
+  face( stepPos );
+  float angle = Util::calcAngFrom( getPos().x, getPos().z, stepPos.x, stepPos.z ) + PI;
+
+  auto x = ( cosf( angle ) * 1.1f );
+  auto y = ( getPos().y + stepPos.y ) * 0.5f; // Get speed from somewhere else?
+  auto z = ( sinf( angle ) * 1.1f );
+
+  Common::FFXIVARR_POSITION3 newPos{ getPos().x + x, y, getPos().z + z };
+  setPos( newPos );
+  setRot( newRot );
+
+  sendPositionUpdate();
+}
+
 bool Sapphire::Entity::BNpc::moveTo( const FFXIVARR_POSITION3& pos )
 {
   if( Util::distance( getPos().x, getPos().y, getPos().z, pos.x, pos.y, pos.z ) <= 4 )
-    // reached destination
+    // Reached destination
     return true;
 
-  float rot = Util::calcAngFrom( getPos().x, getPos().z, pos.x, pos.z );
+  if( m_naviTarget.x == pos.x && m_naviTarget.y == pos.y && m_naviTarget.z == pos.z )
+    // Targets are the same
+    return false;
+
+  // Check if we have to recalculate
+  if( Util::getTimeMs() - m_naviLastUpdate > 500 )
+  {
+    auto pNaviMgr = m_pFw->get< World::Manager::NaviMgr >();
+    auto pNaviProvider = pNaviMgr->getNaviProvider( m_pCurrentZone->getInternalName() );
+
+    if(!pNaviProvider)
+    {
+      Logger::error( "No NaviProvider for zone#{0} - {1}", m_pCurrentZone->getGuId(), m_pCurrentZone->getInternalName() );
+      return false;
```

Comment:
for now maybe we should log an error but continue with pathing to random pos around actor instead of not moving at all if navmesh is unavailable

Reply:
how would you path with no navmesh? just do the naive pathing we had before with the interp between bnpc/player?

Reply:
ya

Reply:
dunno, might just be better to scream and yell about there being no nav provider instead of using that

there seems to be an issue at least from when I was playing with it where bnpcs would bug out and go under the map with zones that have a large y difference